diff --git a/lib/std/http/Client.zig b/lib/std/http/Client.zig
--- a/lib/std/http/Client.zig
+++ b/lib/std/http/Client.zig
@@ -75,7 +75,8 @@
     pub const Criteria = struct {
         host: HostName,
         port: u16,
         protocol: Protocol,
+        proxied: bool,
     };

     /// Finds and acquires a connection from the connection pool matching the criteria.
@@ -90,6 +91,7 @@
             const connection: *Connection = @alignCast(@fieldParentPtr("pool_node", node));
             if (connection.protocol != criteria.protocol) continue;
             if (connection.port != criteria.port) continue;
+            if (connection.proxied != criteria.proxied) continue;

             // Domain names are case-insensitive (RFC 5890, Section 2.3.2.4)
             if (!connection.host().eql(criteria.host)) continue;
@@ -1430,6 +1432,7 @@
     host: HostName,
     port: u16,
     protocol: Protocol,
+    proxied: bool = false,

     proxied_host: ?HostName = null,
     proxied_port: ?u16 = null,
@@ -1449,6 +1452,7 @@
         .host = proxied_host,
         .port = proxied_port,
         .protocol = protocol,
+        .proxied = options.proxied,
     })) |conn| return conn;

     var stream = try host.connect(io, port, .{ .mode = .stream });
@@ -1465,6 +1469,7 @@
                 else => return error.TlsInitializationFailed,
             };
             errdefer tc.destroy();
+            tc.connection.proxied = options.proxied;
             try client.connection_pool.addUsed(io, &tc.connection);
             return &tc.connection;
         },
@@ -1472,6 +1477,7 @@
         .plain => {
             const pc = try Connection.Plain.create(client, proxied_host, proxied_port, stream);
             errdefer pc.destroy();
+            pc.connection.proxied = options.proxied;
             try client.connection_pool.addUsed(io, &pc.connection);
             return &pc.connection;
         },
@@ -1488,6 +1494,7 @@
         .host = path,
         .port = 0,
         .protocol = .plain,
+        .proxied = false,
     })) |node|
         return node;

@@ -1530,6 +1537,7 @@
         .host = proxied_host,
         .port = proxied_port,
         .protocol = proxy.protocol,
+        .proxied = false,
     })) |node| return node;

     var maybe_valid = false;
@@ -1616,9 +1624,12 @@
     }

     // fall back to using the proxy as a normal http proxy
-    const connection = try client.connectTcp(proxy.host, proxy.port, proxy.protocol);
-    connection.proxied = true;
-    return connection;
+    return client.connectTcpOptions(.{
+        .host = proxy.host,
+        .port = proxy.port,
+        .protocol = proxy.protocol,
+        .proxied = true,
+    });
 }

 pub const RequestError = ConnectTcpError || error{
