diff --git a/lib/std/crypto/tls/Client.zig b/lib/std/crypto/tls/Client.zig
--- a/lib/std/crypto/tls/Client.zig
+++ b/lib/std/crypto/tls/Client.zig
@@ -1265,6 +1265,7 @@ fn readIndirect(c: *Client) Reader.Error!usize {
                     .new_session_ticket => {
                         // This client implementation ignores new session tickets.
                     },
                     .key_update => {
+                        if (handshake.len != 1) return failRead(c, error.TlsDecodeError);
                         switch (c.application_cipher) {
                             inline else => |*p| {
                                 const pv = &p.tls_1_3;
